File: sir_ddft_py/examples/sir_ddft_2d.py
#!/usr/bin/env python3

import numpy as np
from numpy import pi, sqrt

import os, os.path
import sir_ddft

from common import run_sim

# Simulation time parameters
DT = 0.1
STEPS = 300

# Spatial simulation parameters (per dimension)
L = 10.0
GRIDPOINTS = 512

# Initial state:
# - Gaussian distribution with variance below and normalized to achieve a given
#   mean population density
# - Initially, 0.1 percent of the total population is infected, the rest is
#   susceptible
VARIANCE = L*L/50
MEAN_DENSITY = 0.3543165399952919

def gaussian(x,y):
    return np.exp(-0.5 / VARIANCE * ((x - L/2)**2 + (y - L/2)**2))
X,Y = np.meshgrid(np.linspace(0,L,GRIDPOINTS), np.linspace(0,L,GRIDPOINTS))
dx = L/(GRIDPOINTS-1)
norm_factor = MEAN_DENSITY / (np.sum(gaussian(X,Y)) * dx * dx / (L * L))
def initfunc(x,y):
    total_pop = norm_factor * gaussian(x,y)
    return [0.999 * total_pop, 0.001 * total_pop, 0.0]

# Initialize model parameters, state and solver
params = sir_ddft.SIRParameters(1.0, 0.1, 0.0)
diff_params = sir_ddft.SIRDiffusionParameters(0.01, 0.01, 0.01)
ddft_params = sir_ddft.SIRDDFTParameters(1.0, 1.0, 1.0, -10, 100, -30, 100)
grid = sir_ddft.Grid2D.new_equidistant(0, L, 0, L, GRIDPOINTS, GRIDPOINTS)
state  = sir_ddft.SIRStateSpatial2D(grid, initfunc)
solver = sir_ddft.SIRDDFT2DSolver(params, diff_params, ddft_params, state, 6)

# Run simulation (see common.py)
run_sim(solver, DT, STEPS, "SIR DDFT model", "sir_ddft_2d", [0,L])

File: sir_ddft/examples/sir_ddft_2d.rs
// Example of the SIR DDFT model in 2D with plotters output

use std::process::ExitCode;

use sir_ddft::ode::{RKF45Solver, ExplicitODESolver};
use sir_ddft::{SIRParameters, SIRDiffusionParameters, SIRDDFTParameters,
    Grid2D, Grid1D, SIRStateSpatial2D, SIRDDFT2DIVP};

mod common;

#[allow(non_snake_case)]
pub fn main() -> ExitCode {
    // Simulation length / resolution parameters
    const DT: f64 = 0.1;
    const STEPS: u64 = 300;
    const GRIDPOINTS: usize = 512;
    const DOMAINSIZE: f64 = 10.0;
    const DX: f64 = DOMAINSIZE / GRIDPOINTS as f64;
    // Other parameters
    const NUM_THREADS: usize = 4;
    // Setup initial state
    let grid = Grid2D::new_cartesian(
        Grid1D::new_equidistant((0.,DOMAINSIZE), GRIDPOINTS),
        Grid1D::new_equidistant((0.,DOMAINSIZE), GRIDPOINTS),
    );
    const VARIANCE: f64 = 50.;
    const MEAN_POPULATION: f64 = 0.3543165399952919;
    let gaussian = |x: f64,y: f64| {
        (-1./(2. * DOMAINSIZE*DOMAINSIZE / VARIANCE) * ((x - DOMAINSIZE/2.).powi(2) + (y - DOMAINSIZE/2.).powi(2))).exp()
    };
    let gaussian_mean = grid.grid()
        .map(|(x,y)| gaussian(x,y))
        .sum::<f64>() * DX * DX / (DOMAINSIZE * DOMAINSIZE);
    let state = SIRStateSpatial2D::new(grid.clone(), |x,y| {
        // Normalize gaussian based on finite domain
        let gaussian_normalized = gaussian(x,y) / gaussian_mean * MEAN_POPULATION;
        (0.999 * gaussian_normalized, 0.001 * gaussian_normalized, 0.0)
    });
    // Setup parameters
    let sir_params = SIRParameters::new(1.0, 0.1, 0.0);
    let diff_params = SIRDiffusionParameters::new(0.03, 0.03, 0.03);
    let ddft_params = SIRDDFTParameters::new(1.0, 1.0, 1.0, -10.0, 100.0, -30.0, 100.0);
    // Create the IVP and solver
    let mut ivp = SIRDDFT2DIVP::new(sir_params, diff_params, ddft_params, state, NUM_THREADS);
    let mut solver = RKF45Solver::<SIRDDFT2DIVP,_>::new();
    // Run solver and collect data for plotting
    let (t,state) = ivp.get_result();
    let mut result = vec![(t, state.to_owned())];
    println!("Running simulation...");
    let progress = indicatif::ProgressBar::new(STEPS);
    for _ in 1..STEPS {
        ivp.add_time(DT);
        solver.integrate(&mut ivp);
        let (t,state) = ivp.get_result();
        result.push((t,state.to_owned()));
        progress.inc(1);
    }
    progress.finish();
    // -- Graphical output --

    if let Err(e) = common::plot(result.as_slice(), Some((0.,DOMAINSIZE)), "SIR-DDFT model in 2D", "sir_ddft_2d.npz") {
        println!("Error: {}", e);
        ExitCode::FAILURE
    }
    else {
        ExitCode::SUCCESS
    }
}